### PR TITLE
Add Dim mode to beat grid lines

### PR DESCRIPTION
**Pull request by williamlemus**
Hi,

I worked on https://github.com/mixxxdj/mixxx/pull/1247 to enable toggling of the beat grid lines. This change is to add a dim mode to the beat grid.

Also, it was mentioned we should move the logic of setting the config file to DlgWaveform. Is moving the check logic from void WaveformWidgetFactory::setDisplayBeatGridAlpha(bool sync) to void DlgPrefWaveform::slotSetGridLinesAlpha(bool alphaGrid) sufficient? Once that's confirmed I can start moving the logic for the other preferences.

### CODE REVIEW

File: src/preferences/dialog/dlgprefwaveform.cpp [hunk 1]
```diff
@@ -236,6 +242,14 @@ void DlgPrefWaveform::slotSetGridLines(bool displayGrid) {
     WaveformWidgetFactory::instance()->setDisplayBeatGrid(displayGrid);
 }
 
+void DlgPrefWaveform::slotSetGridLinesAlpha(bool alphaGrid) {
+    if (m_pConfig) {
+        m_pConfig->set(ConfigKey("[Waveform]", "dimModeCheckBox"), ConfigValue(alphaGrid));
```
Comment (daschuer) [inline src/preferences/dialog/dlgprefwaveform.cpp]: you can use here the template implementation 
m_pConfig->setValue(ConfigKey("[Waveform]", "dimModeCheckBox"), alphaGrid);

File: src/waveform/waveformwidgetfactory.cpp [hunk 2]
```diff
@@ -227,6 +228,9 @@ bool WaveformWidgetFactory::setConfig(UserSettingsPointer config) {
     bool showBeatGrid = m_config->getValue(ConfigKey("[Waveform]", "beatGridLinesCheckBox"), m_beatGridEnabled);
     setDisplayBeatGrid(showBeatGrid);
 
+    bool showBeatGridAlpha = m_config->getValue(ConfigKey("[Waveform]", "dimModeCheckBox"), m_beatGridAlpha);
```
Comment (daschuer) [inline src/waveform/waveformwidgetfactory.cpp]: The parameter should called what it effects, not how it is set. 
"semiTransparentBeatGrid true / false" or "beatGridAlpha 0.3 / 0.9"
In the later case you can still use a checkbox to set one of these values. If we later deside to make the alpha value a skin parameter, the first one works better. The later allows to control the alpha by a slider.

File: src/preferences/dialog/dlgprefwaveform.cpp [hunk 3]
```diff
@@ -69,6 +69,8 @@ DlgPrefWaveform::DlgPrefWaveform(QWidget* pParent, MixxxMainWindow* pMixxx,
             this, SLOT(slotSetNormalizeOverview(bool)));
     connect(beatGridLinesCheckBox, SIGNAL(toggled(bool)),
             this, SLOT(slotSetGridLines(bool)));
+    connect(dimModeCheckBox, SIGNAL(toggled(bool)),
```
Comment (daschuer) [inline src/preferences/dialog/dlgprefwaveform.cpp]: beatGridSemiTransparentCheckBox, or a reasonable shorter name

File: src/waveform/renderers/waveformrenderbeat.cpp [hunk 4]
```diff
@@ -22,9 +22,9 @@ WaveformRenderBeat::~WaveformRenderBeat() {
 void WaveformRenderBeat::setup(const QDomNode& node, const SkinContext& context) {
     m_beatColor.setNamedColor(context.selectString(node, "BeatColor"));
     m_beatColor = WSkinColor::getCorrectColor(m_beatColor).toRgb();
-
     if (m_beatColor.alphaF() > 0.99)
         m_beatColor.setAlphaF(0.9);
```
Comment (daschuer) [inline src/waveform/renderers/waveformrenderbeat.cpp]: This gets overridden later. 
Should we still allow to set an beat grid alpha from the skins? 
... probably not.

File: src/waveform/renderers/waveformwidgetrenderer.cpp [hunk 5]
```diff
@@ -279,6 +280,10 @@ void WaveformWidgetRenderer::setDisplayBeatGrid(bool set) {
     m_enableBeatGrid = set;
 }
 
+void WaveformWidgetRenderer::setDisplayBeatGridAlpha(bool set) {
```
Comment (daschuer) [inline src/waveform/renderers/waveformwidgetrenderer.cpp]: the function name sounds as if it takes the alpha value.

File: src/waveform/waveformwidgetfactory.cpp [hunk 6]
```diff
@@ -453,6 +458,19 @@ void WaveformWidgetFactory::setDisplayBeatGrid(bool sync) {
 
 }
 
+
+void WaveformWidgetFactory::setDisplayBeatGridAlpha(bool sync) {
```
Comment (daschuer) [inline src/waveform/waveformwidgetfactory.cpp]: why sync?

File: src/waveform/renderers/waveformrenderbeat.cpp (hunk 4 above)
Comment (Be-ing) [reply src/waveform/renderers/waveformrenderbeat.cpp]: No, this should be configurable by the user independently of the skin.

File: src/preferences/dialog/dlgprefwaveform.cpp (hunk 3 above)
Comment (Be-ing) [reply src/preferences/dialog/dlgprefwaveform.cpp]: Should we have a slider to let the user select their own desired opacity for the lines?

**Comment by uklotzde** [comment]
Different terms like "dim" and "alpha" are used for the same feature, even in different combinations for variables and function names. Both are not self-explanatory, "dim" is imprecise and "alpha" is too technical. 

I think "semi-transparent beat grid" like @daschuer suggested is a good choice for naming the feature. This phrase can be used as a starting point for consistently deriving all variable and function names.

**Comment by Be-ing** [comment]
> "semi-transparent beat grid"

I think "translucent" is better.

**Comment by daschuer** [comment]
I did not know the word translucent before so I had to look it up and found this: 
https://en.m.wikipedia.org/wiki/Transparency_and_translucency
Since we add no blurry, transparency should be the right term.

**Comment by williamlemus** [comment]
I'll go over and modify the variable/function names to be transparent. @daschuer @Be-ing  Should we make 3 radio buttons or do we make a slider?

**Comment by daschuer** [comment]
For now I am pro radio buttons, the internal interface should allow a slider.

File: src/preferences/dialog/dlgprefwaveform.cpp (hunk 3 above)
Comment (poelzi) [reply src/preferences/dialog/dlgprefwaveform.cpp]: I'm for a slider. Just let the opaque value be a float between 0.0-1.0 and a slider for setting it. Default would then be 0.9.

**Comment by Be-ing** [comment]
There is now a merge conflict in  src/preferences/dialog/dlgprefwaveformdlg.ui

**Comment by Be-ing** [comment]
Ping @williamlemus. It would be nice to have this feature in Mixxx 2.2. Could you fix the merge conflicts in this branch?

**Comment by williamlemus** [comment]
@Be-ing Sorry I'll pick this up again. I see that the target is Sept 30th.

**Comment by Be-ing** [comment]
Ping @williamlemus. We have (tentatively) set the feature freeze target for 2.2 at the end of June. Do you think you'll be able to finish this in the next few weeks?

**Comment by williamlemus** [comment]
yeah I will work on it and get something in before then

**Comment by williamlemus** [comment]
@Be-ing what would be the best way of resolving the conflict?

**Comment by Be-ing** [comment]
Assuming you have [configured an upstream remote](https://mixxx.org/wiki/doku.php/using_git#set_up_git):
```
git fetch upstream
git merge upstream/master
```
The Pro Git ebook has more [details about resolving merge conflicts](https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging#_basic_merge_conflicts).

I just noticed that you are working on the master branch for this PR. For now that is fine, but for the future, make a new branch from master for every pull request as described on the wiki page linked above.

**Comment by williamlemus** [comment]
@Be-ing it was more that it can be tough to edit the xml. I took the upstream and re-added my changes on top.

I'll make a PR from my feature branch going forward. Also, should I be making PR's to the master branch or is there another branch I should be using?

File: src/preferences/dialog/dlgprefwaveformdlg.ui [hunk 7]
```diff
@@ -473,20 +473,36 @@ Select from different types of displays for the waveform, which differ primarily
...
+        <string>Set amount of opacity on beat grid lines.</string>
        </property>
-       <property name="text">
-        <string>Show Beat Markers</string>
+       <property name="maximum">
+        <number>100</number>
+       </property>
+       <property name="value">
+        <number>90</number>
        </property>
```
Comment (Be-ing) [inline src/preferences/dialog/dlgprefwaveformdlg.ui]: It's not totally clear what the units of this number are. Add this to show a percent sign in the spinbox:
```
<property name="suffix">
 <string notr="true"> %</string>
</property>
```

File: src/preferences/dialog/dlgprefwaveformdlg.ui [hunk 8]
```diff
@@ -473,20 +473,36 @@ Select from different types of displays for the waveform, which differ primarily
         <string/>
        </property>
        <property name="text">
-        <string>Beat Grid</string>
+        <string>Beat Grid Opacity</string>
```
Comment (Be-ing) [inline src/preferences/dialog/dlgprefwaveformdlg.ui]: Only capitalize the first word to match the rest of the dialog: "Beat grid opacity"

File: src/preferences/dialog/dlgprefwaveformdlg.ui (hunk 7 above)
Comment (Be-ing) [reply src/preferences/dialog/dlgprefwaveformdlg.ui]: While you're at it, could you also add this for the framerate spinbox:
```
<property name="suffix">
 <string notr="true"> fps</string<
</property>
```

**Comment by Be-ing** [comment]
Aside from those small changes for the .ui file, this looks good to me. :+1:

Would you be interested in finishing #1520 which was abandoned? If you're not interested or don't have time that is fine. I'll finish it in that case.

**Comment by williamlemus** [comment]
I'll take a look at it and let you know in the next couple of days if I can pick it up(if you don't finish it before then).

**Comment by williamlemus** [comment]
How do I go about taking #1520 ?

**Comment by Be-ing** [comment]
If you have your upstream remote configured as described on the [wiki](https://mixxx.org/wiki/doku.php/using_git#set_up_git):
```
git fetch upstream
git checkout upstream/pr/1520
git checkout -b your_new_branch_name # creates a new Git branch
```
Then make the code changes, commit them, and open a new pull request.

**Comment by Be-ing** [comment]
Thanks for adding this nice feature!

**Pull request merged**